src/rgw/rgw_role.cc

```diff
@@ -0,0 +1,259 @@
+#include <errno.h>
+#include <ctime>
+
+#include "common/errno.h"
+#include "common/Formatter.h"
+#include "common/ceph_json.h"
+#include "common/ceph_time.h"
+#include "rgw_rados.h"
+
+#include "include/types.h"
+#include "rgw_string.h"
+
+#include "rgw_common.h"
+#include "rgw_tools.h"
+#include "rgw_role.h"
+
+#define dout_subsys ceph_subsys_rgw
+
+using namespace std;
+
+const string RGWRole::role_name_oid_prefix = "role_names.";
+const string RGWRole::role_oid_prefix = "roles.";
+
+int RGWRole::store_info(bool exclusive)
+{
+  string oid = get_info_oid_prefix() + id;
+
+  bufferlist bl;
+  ::encode(*this, bl);
+  return rgw_put_system_obj(store, store->get_zone_params().roles_pool, oid,
+                bl.c_str(), bl.length(), exclusive, NULL, real_time(), NULL);
+}
+
+int RGWRole::store_name(bool exclusive)
+{
+  RGWNameToId nameToId;
+  nameToId.obj_id = id;
+
+  string oid = get_names_oid_prefix() + name;
+
+  bufferlist bl;
+  ::encode(nameToId, bl);
+  return rgw_put_system_obj(store, store->get_zone_params().roles_pool, oid,
+              bl.c_str(), bl.length(), exclusive, NULL, real_time(), NULL);
+}
+
+int RGWRole::create(bool exclusive)
+{
+  int ret;
+
+  /* check to see the name is not used */
+  ret = read_id(name, id);
+  if (exclusive && ret == 0) {
+    ldout(cct, 0) << "ERROR: name " << name << " already in use for role id "
+                    << id << dendl;
+    return -EEXIST;
+  } else if ( ret < 0 && ret != -ENOENT) {
+    ldout(cct, 0) << "failed reading role id  " << id << ": "
+                  << cpp_strerror(-ret) << dendl;
+    return ret;
+  }
+
+  /* create unique id */
+  uuid_d new_uuid;
+  char uuid_str[37];
+  new_uuid.generate_random();
+  new_uuid.print(uuid_str);
+  id = uuid_str;
+
+  //arn
+  arn = "arn:aws:iam::role" + path + name;
```

Comment:
I think it's mostly fine. One thing Marcus and I had talked about off and on was whether we wanted to use the same namespaces/identifiers as Amazon. (e.g. whether we wanted to use arn:aws: as our namespace prefix.)

I had been leaning toward keeping all our identifiers the same since it improves compatibility and makes copying policies from one place to the other much easier.

The main argument against is that we aren't _actually_ Amazon. I know Marcus at one point suggested making things like that configurable. I'm not sure if there's much benefit to that. (If we have a single IDP backing both AWS STS and Ceph RGW's STS having them both use different ARNs seems a huge negative.)

Anyway, this is a ramble. I think what you have is perfectly fine, but we might want to discuss it tomorrow on the RGW standup just to air the idea out and get feedback in case there's something I'm not thinking of.